### DISCUSSION TITLE
Baking spherical harmonic light volumes with mitsuba

### DISCUSSION THREAD

**Discussion by expenses** [mitsuba-renderer/mitsuba3#412] [Show and tell]
Hi! I've been writing a script for baking out light volumes with mitsuba recently!

https://github.com/expenses/mitsuba-lightvol-baking

The goal was to use it as a global illumination solution something like what unreal engine does: https://docs.unrealengine.com/5.1/en-US/volumetric-lightmaps-in-unreal-engine/.

![20221129_17h14m10s_grim](https://user-images.githubusercontent.com/13566135/204582919-ebd2b4f6-15fb-4bfb-9b8c-f2bd9f6e911e.png)

_ignore the background here :P_

The script works by rendering a whole bunches of vertical slices, with each one containing the 4 sets of coefficients:

![20221129_17h05m00s_grim](https://user-images.githubusercontent.com/13566135/204580761-1df3e6fc-98b9-4e22-8e35-76d404093bf3.png)

I'd love to get some feedback on the script as I'm not quite sure I'm doing a few things correctly:

- I'm using `xargs` to parallelise the rendering, e.g.: `seq 0 24 | xargs -P 8 -I % python script.py models/cbox.xml % -n 25 25 25 --scale 3 3 9 --center 0 1 0`. This is because the sampling loop doesn't really seem to work on multiple cores (with the `llvm_ad_rgb` backend). Is there a good way to fix this?
- The sampling is a bit wierd, as I'm trying to take advantage of both importance sampling but also pathtracing, so I split samples between `scene.sample_emitter_direction` and sampling the pathtracer via `integrator.sample`. Because I'd get fireflies if the pathtracer hit an emitter, I do an initial intersection test to mask things. Ideally `integrator.sample` would return something that I can check if the primary ray hit an emitter and just mask the result based on that, which would mean that I didn't need to trace the same ray twice.
- I'd move to be able to write directly to a 3d texture of some kind (ideally ktx2), I'm not sure what libraries are available for that.

**Comment by Speierers** [comment]
The reason for the low utilization of the multiple cores in your script is that you are parallelizing over the number of probes, and then looping over the number of samples. It would be more efficient to generate a wavefront of size `dr.prod(probe_count) * sample_count` and not use the `dr.Loop`. 

Similarly, in our rendering code, we often generate a wavefront that traces all samples for all pixels at the same time.

Regarding your second point, I am not sure to understand the issue here. The path tracer should be able to converge nicely if you run it with enough samples..

For 3D textures, you should take a look at the `mi.Texture3f` class.

**Reply by expenses** [reply]
This was super helpful, thanks! I didn't really get how wavefronts worked in Mitsuba until reading this.
